Expaning JSONRPCDaemon to Include All RPC Commands

jeffro256 opened the pull request:
Changes:
    1. Added a JSONRPCDaemon method for every Monero Daemon RPC Command that is not obsolete or binary.
    2. Added a restricted() method to JSONRPCDaemon which returns whether the node is in restricted mode.
    3. Added the ability to not pass any POST data to raw_request().
    4. Added RestrictedRPC exception to monero/backends/jsonrpc/exceptions.py
    5. Added a DaemonIsBusy exception to monero/exceptions.py

Motivation: I wanted to add comprehensive daemon RPC functionality to JSONRPCDaemon, which right now only
supports a handful of all the available commands. While it is possible to perform any RPC command with
raw_request() and raw_jsonrpc_request(), I wanted the user to able to know which commands are available,
what inputs the command takes, what output it gives, and leverage the command's power while knowing the
least amount of implementaton details as possible. The Monero Daemon RPC commands do not necessarily have
the most consistent and simple interface, and that is what this commit brings to the table. The new
methods also have the benefit of local input validation, which allows for fewer, more easily decipherable
bugs for the user.

Note: The vast majority of the lines in this commit are documentation. I tried to be as comprehensive as
possible.

jeffro256 commented:
I have started writing tests and will probably finish today, but I wanted to hear your feedback on the current code.

emesik commented on monero/backends/jsonrpc/daemon.py:
```diff
@@ -201,14 +198,1334 @@ def raw_jsonrpc_request(self, method, params=None):
+
+    def _is_valid_256_hex(self, hexhash):
+        try:
+            bytes.fromhex(hexhash)
```
This doesn't work with Python 2.x, where `bytes` is `str` and doesn't support `.fromhex`. I guess `bytearray` would do it.

jeffro256 replied:
Okay! I will include that in the next commit.

emesik reviewed (changes requested):
It seems OK. W must remember however, that when those methods are eventually used by higher level code, the output would change and return object instances instead of raw JSON output.

This should be mentioned in the docs/

jeffro256 commented on monero/backends/jsonrpc/daemon.py:
```diff
@@ -587,68 +587,85 @@ def hard_fork_info(self):
-                if len(host.split('.')) != 4:
-                    raise ValueError('host "{}" is incorrectly formatted'.format(host))
-                for ip_part in host.split('.'):
+        def is_valid_ip(ip):
```
I thought about using the ipaddress module, but I didn't because it isn't standard in < Python 3.3, and I didn't want to add a dependency unless necessary. It would be easy enough to add it though, and would definitely make the code cleaner. It's up to you.

emesik replied:
I think additional dependency for Python2 won't do any harm. It seems a compatible module is available: https://pypi.org/project/py2-ipaddress/

I've already split `test_requirements*.txt` between Py2 and Py3. The same method could be used to make separate `requirements.txt`.

jeffro256 replied:
We might not even need to split it (yet). [This pip package](https://pypi.org/project/ipaddress/) is the official package I'm pretty sure, and it works for all relevant Python versions.

jeffro256 commented:
Yes, my plan was that eventually we could add in the `transactions` for `get_transactions()` equivalent for the different JSON RPC methods, but I wanted to at least have a barebones, special-type-free implementation for those that wanted it. This would have been useful for me, for example, when I wanted to get one-time output information that was in the `get_transactions()` RPC method, but wasn't contained in the `Transaction` class.

jeffro256 commented:
I have uploaded part of the test cases. One change I forgot in the commmit message is that I added `ipaddress` in requirements.txt. I also changed the `bytes.fromhex` line to `bytearray.fromhex` for backwards compatibility.

emesik commented:
No hurry here. I'll be away for at least a week and cannot work on that now. See you next Thursday or even a bit later.

jeffro256 commented:
I have added a new section in the docs, but I'm not very familiar with ReStructured Text so it may need editing. If you can add multiple API references to that doc file, I think it may be good to add the JSONRPCDaemon API reference.

jeffro256 commented:
Any updates?

emesik commented:
Sorry for delaying this so much. A lot of personal stuff happening right now. I'll gather all your PRs and merge them this week.

jeffro256 commented:
Oh no worries, there is no rush! I was just thinking that the notifications weren't going through or something. I wish you luck with your personal stuff

emesik commented:
Merged. Thank you!
